src/minidriver/minidriver.c

```diff
@@ -6477,8 +6477,11 @@ DWORD WINAPI CardGetProperty(__in PCARD_DATA pCardData,
 
 				if (dwFlags == ROLE_USER)
 					p->PinPurpose = PrimaryCardPin;
-				else if (dwFlags == MD_ROLE_USER_SIGN)
+				else if (dwFlags == MD_ROLE_USER_SIGN) {
 					p->PinPurpose = DigitalSignaturePin;
+					/* may need to check pin auto_info for auth_method SC_AC_CONTEXT_SPECIFIC */
+					p->PinCachePolicy.PinCachePolicyType = PinCacheAlwaysPrompt;
```

Comment:
Won't this force PIN re-prompts for every `MD_ROLE_USER_SIGN` PIN, not only the new PIV type?

Minidriver spec says that:
> Note: Windows logon may not work properly if a PIN is not cached. This behavior is by design. 
> Therefore, careful consideration should be given when setting a PIN cache mode to any value other than `PinCacheNormal`.

I guess it really should be a special PKCS15 API PIN flag that says that this PIN needs re-sending before each operation. Or some other way to retrieve this information about PIN from the particular OpenSC card driver.

And then when this OpenSC API PIN flag is set the Minidriver can return `PinCacheAlwaysPrompt` for this PIN.

Reply:
> Won't this force PIN re-prompts for every MD_ROLE_USER_SIGN PIN, not only the new PIV type?

Maybe I have taken a different approach with the  force push today: 

       minidriver.c - Create shadow  MD_ROLE_USER_ALWAYS and MD_ROLE_USER_SIGN_ALWAYS
    
       PKCS11 defines CKA_ALWAYS_AUTHENTICATE attribute for private keys, which is converted
       to `user_consent` for PKCS15 for private keys. Windows has `PinCacheAlwaysPrompt`
       on PINS to accomplish the same thing - prompt user before using a key.
    
       But a key "is secured by" only one pin, but a pin may secure multiple keys where only
       a subset of keys need `PinCacheAlwaysPrompt`
    
       Two new pin roles are defined and use a sc_pkcs15_id auth_id of
       auth_id_md_role_user_always = {"MD_ROLE_USER_ALWAYS", 18};
       and
       auth_id_md_role_user_sign_always = {"MD_ROLE_USER_SIGN_ALWAYS", 18};
    
       When building containers for pkcs15 key objects, if user_consent > 0, the auth_id
       is set to one of the above. Thus when Windows CSP selects a key, it will find
       a pin that matches the correct role for the key.

This is still being tested. But should work for any card the has PKCS15 `user_consent > 0` for 
the User key and/or the Sign key.